ERROR: Cannot find io-package.json in null

Wollte so eben die Version 4.0.2 installieren. Jedoch wurde die Installation mit Fehler abgebrochen: 

Log:

$ ./iobroker upgrade javascript
Update javascript from @4.0.1 to @4.0.2
NPM version: 6.4.1
npm install iobroker.javascript@4.0.2 --production --save --prefix "/opt/iobroker" (System call)
npm
 ERR! 
code
 ETARGETnpm ERR! notarget No matching version found for @types/node@11
npm
 ERR!
 notarget
 In most cases you or one of your dependencies are requestingnpm 
ERR!
 notarget a package version that doesn't exist.
npm ERR! A complete log of this run can be found in:
npm 
ERR!     /root/.npm/_logs/2018-12-05T08_15_54_641Z-debug.log
npm WARN react-beautiful-dnd@9.0.2 requires a peer of react@^16.3.1 but none is installed. You must install peer dependencies yourself.npm WARN react-motion@0.5.2 requires a peer of react@^0.14.9 || ^15.3.0 || ^16.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN react-redux@5.0.7 requires a peer of react@^0.14.0 || ^15.0.0-0 || ^16.0.0-0 but none is installed. You must install peer dependencies yourself.
npm WARN react-beautiful-dnd@8.0.7 requires a peer of react@^16.3.1 but none is installed. You must install peer dependencies yourself.
npm WARN react-motion@0.5.2 requires a peer of react@^0.14.9 || ^15.3.0 || ^16.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN react-redux@5.0.7 requires a peer of react@^0.14.0 || ^15.0.0-0 || ^16.0.0-0 but none is installed. You must install peer dependencies yourself.


npm ERR! code ELIFECYCLEnpm ERR! errno 1
npm ERR! iobroker.javascript@4.0.2 postinstall: `node ./install/installTypings.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the iobroker.javascript@4.0.2 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2018-12-05T08_16_04_984Z-debug.log

ERROR: Cannot find io-package.json in null
ERROR: process exited with code 10

--- a/.travis.yml
+++ b/.travis.yml
@@ -3,7 +3,6 @@ os:
   - osx
 language: node_js
 node_js:
-  - '4'
   - '6'
   - '8'
   - '10'

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -1,7 +1,6 @@
 version: 'test-{build}'
 environment:
   matrix:
-    - nodejs_version: '4'
     - nodejs_version: '6'
     - nodejs_version: '8'
     - nodejs_version: '10'

--- a/install/installTypings.js
+++ b/install/installTypings.js
@@ -3,20 +3,20 @@
 // The adapter includes typings for the lowest supported NodeJS version.
 // This script updates them to the installed version
 
-const { spawn } = require("child_process");
-const semver = require("semver");
+const { spawn } = require('child_process');
+const semver = require('semver');
 const installedNodeVersion = semver.coerce(process.versions.node).major;
 
 function fail(reason) {
-    console.error("Could not install NodeJS typings. This is not critical.");
-    console.error("Reason: \n" + reason);
+    console.error('Could not install NodeJS typings. This is not critical.');
+    console.error('Reason: \n' + reason);
     // This is not critical!
     process.exit(0);
 }
 
 // Find latest version on npm
 console.log('Installing NodeJS typings...');
-npmCommand("view", ["@types/node", "version"], {stdout: "pipe", stderr: "pipe"})
+npmCommand('view', ['@types/node', 'version'], {stdout: 'pipe', stderr: 'pipe'})
     .then(cmdResult => {
         if (cmdResult.exitCode !== 0) {
             return fail(cmdResult.stderr);
@@ -24,13 +24,13 @@ npmCommand("view", ["@types/node", "version"], {stdout: "pipe", stderr: "pipe"})
         const latestVersion = semver.coerce(cmdResult.stdout).major;
         console.log(`latest @types: ${latestVersion}, installed node: ${installedNodeVersion}`);
         return semver.gt(`${installedNodeVersion}.0.0`, `${latestVersion}.0.0`)
-            ? "latest" // The installed version is too new, install latest
+            ? 'latest' // The installed version is too new, install latest
             : installedNodeVersion.toString()
         ;
     })
     .then(targetVersion => {
         // Install the desired version
-        return npmCommand("i", [`@types/node@${targetVersion}`], {stdout: "ignore", stderr: "pipe"});
+        return npmCommand('i', [`@types/node@${targetVersion}`], {stdout: 'ignore', stderr: 'pipe'});
     })
     .then(cmdResult => {
         if (cmdResult.exitCode !== 0) {
@@ -38,7 +38,7 @@ npmCommand("view", ["@types/node", "version"], {stdout: "pipe", stderr: "pipe"})
         } else {
             process.exit(0);
         }
-    })
+    });
 
 // TODO: the following code is copied from a js-controller fork
 // It should be moved to the core and referenced from there in a future version
@@ -96,7 +96,7 @@ function npmCommand(command, npmArgs, options) {
             /** @type {string} */
             let bufferedStderr;
             const cmd = spawn(npmBinary, [command].concat(npmArgs), spawnOptions)
-                .on("close", (code, signal) => {
+                .on('close', (code, signal) => {
                     resolve({
                         exitCode: code,
                         signal,
@@ -106,21 +106,21 @@ function npmCommand(command, npmArgs, options) {
                 });
             // Capture stdout/stderr if requested
             if (options.stdout === 'pipe') {
-                bufferedStdout = "";
-                cmd.stdout.on("data", chunk => {
+                bufferedStdout = '';
+                cmd.stdout.on('data', chunk => {
                     const buffer = Buffer.isBuffer(chunk)
                         ? chunk
-                        : new Buffer(chunk, "utf8")
+                        : new Buffer(chunk, 'utf8')
                         ;
                     bufferedStdout += buffer;
                 });
             }
             if (options.stderr === 'pipe') {
-                bufferedStderr = "";
-                cmd.stderr.on("data", chunk => {
+                bufferedStderr = '';
+                cmd.stderr.on('data', chunk => {
                     const buffer = Buffer.isBuffer(chunk)
                         ? chunk
-                        : new Buffer(chunk, "utf8")
+                        : new Buffer(chunk, 'utf8')
                         ;
                     bufferedStderr += buffer;
                 });

--- a/install/installTypings.js
+++ b/install/installTypings.js
@@ -3,22 +3,130 @@
 // The adapter includes typings for the lowest supported NodeJS version.
 // This script updates them to the installed version
 
-let nodeVersion;
-try {
-    nodeVersion = process.versions.node.split('.')[0];
-} catch (e) {
-    console.error('Could not determine NodeJS version, installing latest typings...');
+const { spawn } = require("child_process");
+const semver = require("semver");
+const installedNodeVersion = semver.coerce(process.versions.node).major;
+
+function fail(reason) {
+    console.error("Could not install NodeJS typings. This is not critical.");
+    console.error("Reason: \n" + reason);
+    // This is not critical!
+    process.exit(0);
 }
 
-if (!nodeVersion) nodeVersion = 'latest';
+// Find latest version on npm
+console.log('Installing NodeJS typings...');
+npmCommand("view", ["@types/node", "version"], {stdout: "pipe", stderr: "pipe"})
+    .then(cmdResult => {
+        if (cmdResult.exitCode !== 0) {
+            return fail(cmdResult.stderr);
+        }
+        const latestVersion = semver.coerce(cmdResult.stdout).major;
+        console.log(`latest @types: ${latestVersion}, installed node: ${installedNodeVersion}`);
+        return semver.gt(`${installedNodeVersion}.0.0`, `${latestVersion}.0.0`)
+            ? "latest" // The installed version is too new, install latest
+            : installedNodeVersion.toString()
+        ;
+    })
+    .then(targetVersion => {
+        // Install the desired version
+        return npmCommand("i", [`@types/node@${targetVersion}`], {stdout: "ignore", stderr: "pipe"});
+    })
+    .then(cmdResult => {
+        if (cmdResult.exitCode !== 0) {
+            return fail(cmdResult.stderr);
+        } else {
+            process.exit(0);
+        }
+    })
 
-console.log('installing NodeJS typings...');
-const spawn = require('child_process').spawn;
-const npmCommand = /^win/.test(process.platform) ? 'npm.cmd' : 'npm';
-// @ts-ignore The last argument is available starting with NodeJS 8
-const install = spawn(npmCommand, ['i', '--save', `@types/node@${nodeVersion}`], {windowsHide: true});
-install.stdout.pipe(process.stdout);
-install.stderr.pipe(process.stderr);
-install.on('close', (code, signal) => {
-    process.exit(code);
-});
+// TODO: the following code is copied from a js-controller fork
+// It should be moved to the core and referenced from there in a future version
+
+/**
+ * @typedef {object} NpmCommandOptions
+ * @property {string} cwd The directory to execute the command in
+ * @property {NodeJS.ReadStream}  stdin  Where to redirect the stdin. Default: process.stdin
+ * @property {NodeJS.WriteStream | "pipe" | "ignore"} stdout A write stream to redirect the stdout, "ignore" to ignore it or "pipe" to return it as a string. Default: process.stdout
+ * @property {NodeJS.WriteStream | "pipe" | "ignore"} stderr A write stream to redirect the stderr, "ignore" to ignore it or "pipe" to return it as a string. Default: process.stderr
+ */
+
+/**
+ * @typedef {object} NpmCommandResult
+ * @property {number} exitCode - The exit code of the spawned process
+ * @property {string?} signal - The signal the process received before termination
+ * @property {string?} stdout - If options.stdout was set to "buffer", this contains the stdout of the spawned process
+ * @property {string?} stderr - If options.stderr was set to "buffer", this contains the stderr of the spawned process
+ */
+
+/**
+ * Executes an npm command (e.g. install) and returns the exit code and (if requested) the stdout
+ * @param {string} command The npm command to execute
+ * @param {string[]} [npmArgs] The command line arguments for the npm command
+ * @param {Partial<NpmCommandOptions>} [options] (optional) Some options for the command execution
+ * @returns {Promise<NpmCommandResult>}
+ */
+function npmCommand(command, npmArgs, options) {
+    if (typeof npmArgs === 'object' && !Array.isArray(npmArgs)) {
+        // no args were given
+        options = npmArgs;
+        npmArgs = undefined;
+    }
+    if (options == null) options = {};
+    if (npmArgs == null) npmArgs = [];
+
+    const npmBinary = /^win/.test(process.platform) ? 'npm.cmd' : 'npm';
+    /** @type {import("child_process").SpawnOptions} */
+    const spawnOptions = {
+        stdio: [
+            options.stdin || process.stdin,
+            options.stdout || process.stdout,
+            options.stderr || process.stderr,
+        ],
+        // @ts-ignore This option exists starting with NodeJS 8
+        windowsHide: true,
+    };
+    if (options.cwd != null) spawnOptions.cwd = options.cwd;
+
+    // Now execute the npm process and avoid throwing errors
+    return new Promise((resolve) => {
+        try {
+            /** @type {string} */
+            let bufferedStdout;
+            /** @type {string} */
+            let bufferedStderr;
+            const cmd = spawn(npmBinary, [command].concat(npmArgs), spawnOptions)
+                .on("close", (code, signal) => {
+                    resolve({
+                        exitCode: code,
+                        signal,
+                        stdout: bufferedStdout,
+                        stderr: bufferedStderr
+                    });
+                });
+            // Capture stdout/stderr if requested
+            if (options.stdout === 'pipe') {
+                bufferedStdout = "";
+                cmd.stdout.on("data", chunk => {
+                    const buffer = Buffer.isBuffer(chunk)
+                        ? chunk
+                        : new Buffer(chunk, "utf8")
+                        ;
+                    bufferedStdout += buffer;
+                });
+            }
+            if (options.stderr === 'pipe') {
+                bufferedStderr = "";
+                cmd.stderr.on("data", chunk => {
+                    const buffer = Buffer.isBuffer(chunk)
+                        ? chunk
+                        : new Buffer(chunk, "utf8")
+                        ;
+                    bufferedStderr += buffer;
+                });
+            }
+        } catch (e) {
+            // doesn't matter, we return the exit code in the "close" handler
+        }
+    });
+}

--- a/lib/tools.js
+++ b/lib/tools.js
@@ -3,11 +3,11 @@
  * @param {any} it The variable to test
  */
 function isObject(it) {
-	// This is necessary because:
-	// typeof null === 'object'
-	// typeof [] === 'object'
-	// [] instanceof Object === true
-	return Object.prototype.toString.call(it) === '[object Object]';
+    // This is necessary because:
+    // typeof null === 'object'
+    // typeof [] === 'object'
+    // [] instanceof Object === true
+    return Object.prototype.toString.call(it) === '[object Object]';
 }
 
 /**
@@ -15,11 +15,11 @@ function isObject(it) {
  * @param {any} it The variable to test
  */
 function isArray(it) {
-	if (Array.isArray != null) return Array.isArray(it);
-	return Object.prototype.toString.call(it) === '[object Array]';
+    if (Array.isArray != null) return Array.isArray(it);
+    return Object.prototype.toString.call(it) === '[object Array]';
 }
 
 module.exports = {
-	isArray,
-	isObject,
+    isArray,
+    isObject,
 };

--- a/package.json
+++ b/package.json
@@ -31,6 +31,7 @@
     "coffee-script": "^1.12.7",
     "node-schedule": "1.3.0",
     "request": "^2.88.0",
+    "semver": "^5.6.0",
     "suncalc": "^1.8.0",
     "typescript": "^3.1.6",
     "virtual-tsc": "^0.4.6",

